gemma-rest/src/main/java/ubic/gemma/rest/analytics/ga4/FromRequestClientIdRetrievalStrategy.java

```diff
@@ -0,0 +1,26 @@
+package ubic.gemma.rest.analytics.ga4;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.UUID;
+
+@Component
+public class FromRequestClientIdRetrievalStrategy implements ClientIdRetrievalStrategy {
+
+    @Autowired
+    private HttpServletRequest request;
+
+    @Override
+    public String get() {
+        String clientId = request.getHeader( "X-Gemma-Client-ID" );
+        if ( StringUtils.isNotBlank( clientId ) && StringUtils.isAlphanumeric( clientId.trim() ) ) {
+            return clientId.trim();
+        } else {
+            // random client ID
+            return UUID.randomUUID().toString();
```

Comment:
Maybe reuse the random client ID strategy? or have a backup strategy chain.

Reply:
We might be able to regroup certain HTTP requests under the same user ID using the session.

Reply:
This would be used for clients that are not passing the `X-Gemma-Client-ID` header.